templates/cmd/main.go.tpl

```diff
@@ -126,5 +126,7 @@ func main() { //nolint: funlen // Why: We can't dwindle this down anymore withou
 
 	if err := async.RunGroup(acts).Run(ctx); err != nil {
 		log.Warn(ctx, "shutting down service", events.NewErrorInfo(err))
+    // If there is an error it should exit with a code to indicate it
+		defer os.Exit(1)
```

Comment:
We shouldn't do this. `defer` is LIFO so this will render all other defers useless. Let me show you the right pattern

Reply:
```go
func main(){
  exitCode := 1
  defer func(){
    os.Exit(exitCode)
  }()
  
  // .. do things, defer things, whatever
  
  // if we make it to the end of func main without erroring and returning we can assume success
  exitCode = 0
}
```

Reply:
That might not fit into what we have right now, but I hope the idea is there and it can be adapted. The important part is that this `defer` that contains the `os.Exit` call needs to be the **first** `defer` statement in `func main`

Reply:
This is heavily obfuscated imperative programming. The exit code is far removed from what triggers it, and you're making readers go back to the start of the function to understand why an assign is non-pointless. Much more readable is something functional:

```golang
func main() {
  exitCode := runMain()
  os.Exit(exitCode)
}

func runMain() int {
  if err != nil {
    // Stuff is bad.
    return 1
  }
  // Stuff is good.
  return 0
}
```

Reply:
While I appreciate your view here and I largely agree with it this is one case that I choose to go against my own values on. This is one of those things you implement exactly once and then forget about forever. It never needs to change and it always will work. Just leave a comment to make sure nobody ever puts code above the defer in func main.

Reply:
If we did go with the pattern you just suggested I'd much rather `runMain` return an error instead of an exit code though

Reply:
I've used both patterns. I like the "bad" one more (the one I suggested) personally. I won't die on this hill though.

Reply:
Returning an error is fine - although in the general case it should include an exit code value, for when you want multiple exit codes in e.g. a CLI.

The other argument against this pattern is it makes `main` untestable, since you've called `os.Exit`. You obviously need this somewhere, and the scope where you lose coverage should be as small as possible.

Reply:
Wouldn't it be a breaking change to have `main` return a code for this?

Reply:
Yes, it's not in-scope for this PR (or for a quick-fix).